MaxText/inference_microbenchmark.py

```diff
@@ -154,36 +148,45 @@ def write_results(results, filename=""):
-  print("\nFor usage in analyze_sharegpt.py :")
-  print(f"PREFILL_BUCKET_SIZE_TO_MS = {prefill_bucket_size_to_ms}")
-  print(f"SYSTEM_TIME_PER_DECODE_TOKEN_MS = {results['AutoRegressive']['ar_step_in_ms_per_seq']}")
+  if "Prefill" in results:
+    prefill_bucket_size_to_ms = {}
+    for k, v in results["Prefill"].items():
+      prefill_bucket_size_to_ms[int(k)] = round(v["prefill_time_in_ms"], 3)
+    print("\nFor usage in analyze_sharegpt.py :")
+    print(f"PREFILL_BUCKET_SIZE_TO_MS = {prefill_bucket_size_to_ms}")
+
+  if "AutoRegressive" in results:
+    print(f"SYSTEM_TIME_PER_DECODE_TOKEN_MS = {results['AutoRegressive']['ar_step_in_ms_per_seq']}")
 
 
 def main(config):
   engine = maxengine.MaxEngine(config)
   params = engine.load_params()
-  prefill_lengths = [64, 128, 256, 512, 1024]
-  benchmark_loop_iters = 10
+  prefill_lengths = [int(l) for l in config.inference_microbenchmark_prefill_lengths.split(",")]
+  stages_to_benchmark = config.inference_microbenchmark_stages.split(",")
+  benchmark_loop_iters = config.inference_microbenchmark_loop_iters
+
   text = config.prompt
   metadata = engine.get_tokenizer()
   vocab = token_utils.load_vocab(metadata.path, metadata.extra_ids)
 
   decode_state = engine.init_decode_state()
-  _, cache_size, _ = summarize_pytree_data(decode_state['cache'], name="Cache")
-  num_model_params, model_size, _ = summarize_pytree_data(params, name="Model")
-
-  benchmark_results = {"Prefill": {}}
-  benchmark_results["AutoRegressive"], decode_state = ar_benchmark(
-    config, engine, params, decode_state, iters=benchmark_loop_iters, cache_size=cache_size, model_size=model_size)
-  for prefill_length in prefill_lengths:
-    tokens, true_length = token_utils.tokenize_and_pad(
-      text, vocab, is_bos=True, prefill_lengths=[prefill_length])
-    benchmark_results["Prefill"][prefill_length], decode_state = prefill_benchmark(
-      config, engine, params, decode_state, tokens, true_length,
-      iters=benchmark_loop_iters, num_model_params=num_model_params)
+  _, cache_size, _ = inference_utils.summarize_pytree_data(decode_state['cache'], name="Cache")
+  num_model_params, model_size, _ = inference_utils.summarize_pytree_data(params, name="Model")
+
+  benchmark_results = {}
+  if "prefill" in stages_to_benchmark:
+    benchmark_results["Prefill"] = {}
+    for prefill_length in prefill_lengths:
+      tokens, true_length = token_utils.tokenize_and_pad(
+        text, vocab, is_bos=True, prefill_lengths=[prefill_length])
+      benchmark_results["Prefill"][prefill_length], decode_state = prefill_benchmark(
+        config, engine, params, decode_state, tokens, true_length,
+        iters=benchmark_loop_iters, num_model_params=num_model_params)
+
+  if "generate" in stages_to_benchmark:
+    benchmark_results["AutoRegressive"], decode_state = ar_benchmark(
+      config, engine, params, decode_state, iters=benchmark_loop_iters, cache_size=cache_size, model_size=model_size)
```

Comment:
For running just generate benchmark, you still need to populate kv cache to produce proper perf numbers?

Reply:
You will still need to initialize a decode_state for generate step calculation